# Code repository for the following study:

Lei Zhou#, Ken HO Yu#, Tin-Lok Wong#, Zhao Zhang, Chun-Ho Chan, Jane HC Loong, No√©lia Che, Hua-Jian Yu, Kel Vin Tan, Man Tong, Elly SW Ngan, Joshua WK Ho, Stephanie Ma. Lineage tracing and single-cell analysis reveal proliferative Prom1+ tumor-propagating cells and their dynamic cellular transition during liver cancer progression. 2021

This repository contains the scripts which produced the bioinformatics figures in the paper, namely Figures 5, S5, 6, S6, 7, S7, and S9. It also contains some processed input data required by the scripts, and the direct outputs of the scripts (the Figures) as well as some intermediate outputs. Processed data required that are too large to be placed here can be downloaded in the links below (Recommended downloads).

The Gene Expression Omnibus (GEO) repository of the study, which stores the UMI counts matrices and raw sequencing reads of the samples, can be accessed through the GEO accession [GSE181515](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE181515)

## Recommended downloads

These are needed to exactly reproduce the figures because it isn't possible to generate the exact same integration result (numbered clusters, UMAP coordinates) using different versions of R, [Seurat](https://satijalab.org/seurat/), their dependencies, and different platforms.

**6_samples_integrated_f500_npcs50_res2.rds**  (to be placed in HCC_Stem_scRNAseq/outs/Figure_5_and_S5/5BEFG_S5AB_integration_and_annotation/): https://drive.google.com/file/d/1swecsh1aoTH-ia5xPKQHZb6PLlfSFpfr/view?usp=sharing

The Seurat integrated object of all cell types of the six scRNA-seq samples

**HCC.final.SCT_pc10_res0.1.rds** (to be placed in HCC_Stem_scRNAseq/outs/Figure_5_and_S5/5BEFG_S5AB_integration_and_annotation/): https://drive.google.com/file/d/1ZTL_456l_wWers5dYiM55CzgdeQgwQSQ/view?usp=sharing

The Seurat re-integrated object of the HCC cells only of the six scRNA-seq samples

**processed_data_for_scVelo** (to be placed in HCC_Stem_scRNAseq/data/Figure_6_and_S6/6E_S6D_scVelo_RNA_velocity/): https://drive.google.com/file/d/1wP328GxieXeWtcsEVsLlt-7_JZm2_d2_/view?usp=sharing

The Seurat object loom file (generated by [loomR](https://github.com/mojaveazure/loomR)) and the spliced/unspliced counts loom file (generated by [velocyto](http://velocyto.org/velocyto.py/index.html)) as input to [scVelo](https://scvelo.readthedocs.io/)

## How to run the scripts

The vast majority of the scripts are in R. On macOS, because the default R working directory is the home directory (/Users/_user_), the repository can simply be downloaded and placed in the home directory and R scripts will run. On other platforms or if the repository is not placed in the home directory, at each script's very beginning a setwd() command or an edit to the 'dir' variable providing the full path to the repository may be needed.

For the Python and Perl scripts, ```cd``` into the script's directory to run them.

## Additional downloads

**Cell_Ranger_filtered_matrices** (to be placed in HCC_Stem_scRNAseq/data/Figure_5_and_S5/5BEFG_S5AB_integration_and_annotation/): https://drive.google.com/file/d/1I5-e8C0Mzm7n5Qk2cTnXazFRNGzH7x7g/view?usp=sharing

The filtered UMI counts matrices of the six samples generated by [Cell Ranger](https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/what-is-cell-ranger), but furthermore with sample IDs appended to the end of each barcode to ensure barcodes are unique across samples. Only needed if attempting to re-run the Seurat integration script.